### PR TITLE
League page

### PR DESCRIPTION
**Pull request by Vexrax**
# DO NOT DEPLOY THIS CHANGE WITHOUT ADDING THE GOOGLE API KEY

### How the page looks (Zoomed out a bit)
image

### Whats in this CR
- the frontend for the league pages
- Automatic loading from the premade sheets 
- Season Model changes to support loading from the sheets
- Placeholder page for the individual team page

### Followups
- create ingress into page: https://github.com/users/afieldi/projects/2/views/1?pane=issue&itemId=33954162
- Create divine and draft league parsers: https://github.com/users/afieldi/projects/2/views/1?pane=issue&itemId=33954164
- Figure out a more efficient way to load teams in : https://github.com/users/afieldi/projects/2/views/1?pane=issue&itemId=33954187
- Change the dragon stat to something more interesting 
- Run a backfill to update the teams
- Hook in the team to saving: https://github.com/users/afieldi/projects/2/views/1?pane=issue&itemId=33971103

### CODE REVIEW

File: BackEnd/scripts/generateTeamModel.ts [hunk 1]
```diff
@@ -29,16 +29,27 @@ type RisenSheetRow = {
 
 async function buildTeam(seasonId: number, displayName: string, abbreviation: string, win: number, loss: number, opgg: string) {
   await ensureConnection();
+  let team = await findTeam(displayName, abbreviation, seasonId);
+
   if(await findTeam(displayName, abbreviation, seasonId)) {
```
Comment (afieldi) [inline BackEnd/scripts/generateTeamModel.ts:34]: Duplicate call here. Just replace with `if(team)`

File: BackEnd/scripts/generateTeamModel.ts [hunk 2]
```diff
@@ -29,16 +29,27 @@ type RisenSheetRow = {
...
+    console.log(`the team: ${displayName} already exists! Updating instead`);
+    team.seasonId = seasonId;
+    team.displayName = displayName;
+    team.win = win;
+    team.loss = loss;
+    team.abbreviation = abbreviation;
+    team.wonSeason = false;
+    let rows : TeamModel[] = [];
+    rows.push(team);
+    await SaveObjects(rows, TeamModel);
```
Comment (afieldi) [inline BackEnd/scripts/generateTeamModel.ts:44]: this can be condensed to `await SaveObjects([team], TeamModel);`

File: BackEnd/src/api/teams.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import { GetTeamsRequest, GetTeamsResponse } from '../../../Common/Interface/Internal/teams';
+import { GetTeamsBySeasonId } from '../business/teams';
+
+const router: Router = express.Router();
+
+router.post('/by-seasonId/:seasonId', async(req: TypedRequest<GetTeamsRequest>, res: TypedResponse<GetTeamsResponse>) => {
+  try {
+    logger.info(`Getting the teams for seasonId: ${req.params.seasonId}`);
+    const seasonId = Number(req.params.seasonId);
+    const response= await GetTeamsBySeasonId(seasonId);
```
Comment (afieldi) [inline BackEnd/src/api/teams.ts:13]: Linter didn't catch this? lol

File: BackEnd/src/business/games.ts [hunk 4]
```diff
@@ -40,6 +41,7 @@ export async function SaveDataByMatchId(matchId: string, updatePlayerStats: bool
   const savedGameModel: GameModel = await SaveSingleMatchById(matchId, gameData);
 
   if (updatePlayerStats) {
+    await buildRisenTeams();
```
Comment (afieldi) [inline BackEnd/src/business/games.ts:44]: Should this be moved to a higher level function? Thoughts?

File: BackEnd/src/business/season.ts [hunk 5]
```diff
@@ -10,3 +11,11 @@ export async function CreateSeason(seasonName: string, providerId: number): Prom
 export async function GetSeasons(active: boolean): Promise<SeasonModel[]> {
   return await GetDbSeasons(active);
 }
+
+export async function GetSeasonBySearchName(searchName: string): Promise<SeasonModel> {
```
Comment (afieldi) [inline BackEnd/src/business/season.ts:15]: Can we get rid of this?

File: BackEnd/src/db/games.ts [hunk 6]
```diff
@@ -408,3 +408,9 @@ export async function GetDbPlayerGamesBySeasonId(seasonId: string): Promise<Play
   }
   return await PlayerGameModel.find(filter);
 }
+
+export async function GetDbGamesBySeasonId(seasonId: number ): Promise<PlayerGameModel[]> {
```
Comment (afieldi) [inline BackEnd/src/db/games.ts:412]: Can this be renamed to `GetDbPlayerGamesBySeasonId`. As there is also a GameModel.

File: front-end/src/App.tsx [hunk 7]
```diff
@@ -29,9 +31,11 @@ function App() {
             <Route path="/" element={<Home />}></Route>
             <Route path="/search" element={<SearchPage />}></Route>
             <Route path="/player/:playerName" element={<Player />}></Route>
+            <Route path="/leagues/:leagueName" element={<LeaguePage />}></Route>
             <Route path="/404" element={<Error404 />}></Route>
             <Route path="/statexport" element={<StatExport />}></Route>
             <Route path="/leaderboard" element={<Leaderboards />}></Route>
+            <Route path="/leagues/:leagueName/:teamName" element={<TeamPage />}></Route>
```
Comment (afieldi) [inline front-end/src/App.tsx:38]: Should this be teamId? Or are we using a searchname for this as well?

Comment (Vexrax) [reply front-end/src/App.tsx:38]: Ill probably update this when i get to implementing it but i think its going to actually be the abbreviation. That way people can just replace in the link with their own team name and go straight to the page

File: BackEnd/scripts/generateTeamModel.ts (hunk 1 above)
Comment (Vexrax) [reply BackEnd/scripts/generateTeamModel.ts:34]: Oh lmao, i fixed this in the actual code but not the script, will fix

File: BackEnd/src/api/teams.ts (hunk 3 above)
Comment (Vexrax) [reply BackEnd/src/api/teams.ts:13]: Weird, fixed.

File: front-end/src/components/league-page/leaderboard/leaderboardCard.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      minWidth: props.width,
+      minHeight: props.height,
+      flexGrow: 1,
+      flexWrap: 'wrap',
+    }}>
+      <Box sx={{ p: 0.2, borderRadius: '4px 4px 0px 0px', background: getGradient(theme.palette.risenBoxBg.main)  }}>
+        {props.title}
+        <hr/>
+      </Box>
+      <Box sx={{ background: getGradient(theme.palette.risenBoxBg.main) }}>
```
Comment (afieldi) [inline front-end/src/components/league-page/leaderboard/leaderboardCard.tsx:32]: Just wondering, why not just use a table for this?

File: BackEnd/src/business/games.ts (hunk 4 above)
Comment (Vexrax) [reply BackEnd/src/business/games.ts:44]: I dont think this should be in a higher level function because its directly related to game saving but im looking at this code again and i realize that i made a mistake. I think this call needs to happen before the games save.

Will update and test

File: front-end/src/components/league-page/leaderboard/leaderboardCard.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  const theme = useTheme() as Theme;
+
+  return (
+    <Box sx={{
+      minWidth: props.width,
+      minHeight: props.height,
+      flexGrow: 1,
+      flexWrap: 'wrap',
+    }}>
+      <Box sx={{ p: 0.2, borderRadius: '4px 4px 0px 0px', background: getGradient(theme.palette.risenBoxBg.main)  }}>
```
Comment (afieldi) [inline front-end/src/components/league-page/leaderboard/leaderboardCard.tsx:28]: I know this is a decent sized refactor, but can basically all of the gradients be moved to the risenBox? And have a parameter like useGradient, so it would be easy to ship to all pages later if we want. This file in particular should then be able to use a RisenBox as the top level box with the gradient I think.

File: front-end/src/components/league-page/leaderboard/row.tsx [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    mainValueElement: React.ReactNode,
+}
+
+export interface RowMainValue {
+  value: number,
+  formatter: (value: number) => string
+}
+export function Row(bgColor: string, nameTypography: React.ReactNode, gamesTypography: React.ReactNode, mainValueTypography: React.ReactNode) {
+  return (
+    <Box sx={{ pt: 0.5, pb: 0.5, display: 'flex', flexDirection: 'row', justifyContent:'space-between', background: getGradient(bgColor) }}>
```
Comment (afieldi) [inline front-end/src/components/league-page/leaderboard/row.tsx:19]: You can set these to be slightly transparent I think, so the parents background gradient will seep through. Instead of needing to manually set every gradient.

File: BackEnd/src/business/season.ts (hunk 5 above)
Comment (Vexrax) [reply BackEnd/src/business/season.ts:15]: No, we need it for the frontend since we use the searchName for the URL
http://localhost:3000/leagues/summer2023unstoppable

File: BackEnd/src/db/games.ts (hunk 6 above)
Comment (Vexrax) [reply BackEnd/src/db/games.ts:412]: That name already exists, ive deleted this function and instead used the existing function

File: front-end/src/components/league-page/leaderboard/leaderboardCard.tsx (hunk 9 above)
Comment (Vexrax) [reply front-end/src/components/league-page/leaderboard/leaderboardCard.tsx:28]: For a lot of the boxes i use in this page the risen box doesnt actually meet the design i was going for so i ended up not using it for a lot of things. I can make a followup task to clean it up but id prefer not to do it in this CR since its already big and i dont really want add 40+ more files when i refactor

File: front-end/src/components/league-page/leaderboard/leaderboardCard.tsx (hunk 8 above)
Comment (Vexrax) [reply front-end/src/components/league-page/leaderboard/leaderboardCard.tsx:32]: Table didnt have the look that i was going for so i just built the component myself.

File: front-end/src/components/league-page/leaderboard/leaderboardCard.tsx (hunk 9 above)
Comment (Vexrax) [reply front-end/src/components/league-page/leaderboard/leaderboardCard.tsx:28]: [****](https://github.com/users/afieldi/projects/2/views/1?pane=issue&itemId=33971381)

File: front-end/src/components/league-page/leaderboard/row.tsx (hunk 10 above)
Comment (Vexrax) [reply front-end/src/components/league-page/leaderboard/row.tsx:19]: Each row needs to manually set the gradiant the the bgColor is not the same. 

If you see this image, the rows are slightly off colored.
image

Since each row is difference i need to calc the gradiant since its a diff color.

Comment (afieldi) [reply front-end/src/components/league-page/leaderboard/row.tsx:19]: Yeah ik. Every other row is the same as the background, then the others are darker. For the darker rows you set the background to a slightly transparent black, and it will work with the background to make the same effect.

File: front-end/src/App.tsx (hunk 7 above)
Comment (afieldi) [reply front-end/src/App.tsx:38]: Make sense.

File: BackEnd/src/db/games.ts (hunk 6 above)
Comment (afieldi) [reply BackEnd/src/db/games.ts]: Ok perfect.

File: front-end/src/components/league-page/leaderboard/leaderboardCard.tsx (hunk 8 above)
Comment (afieldi) [reply front-end/src/components/league-page/leaderboard/leaderboardCard.tsx:32]: Ok, ideally this can be moved out into its own full component so it can be reused then

**Review by afieldi — APPROVED** [review]
